core/state/state_object.go

```diff
@@ -268,8 +289,10 @@ func (s *stateObject) finalise(prefetch bool) {
 			delete(s.pendingStorage, key)
 		}
 	}
-	if s.db.prefetcher != nil && prefetch && len(slotsToPrefetch) > 0 && s.data.Root != types.EmptyRootHash {
-		s.db.prefetcher.prefetch(s.addrHash, s.data.Root, s.address, slotsToPrefetch)
+	if s.db.prefetcher != nil && len(slotsToPrefetch) > 0 && s.data.Root != types.EmptyRootHash {
+		if err := s.db.prefetcher.prefetch(s.addrHash, s.data.Root, s.address, slotsToPrefetch); err != nil {
```

Comment:
It could be problematic.

In `updateRoot`, it will:
- flush the pending storage changes into the hasher(trie)
- compute the new storage root

at that time, prefetcher is already terminated. Unfortunately, it might be possible that there are still some cached storage changes, e.g. caused by some rules post-execution (e.g. withdrawal, although withdrawal won't touch storages, but we might have more post-execution rules in the future)

Reply:
EDIT: all the `finalization` has been done in `IntermediateRoot` before the prefetcher termination. So the issue is not relevant anymore.


But we should remove `	s.finalise` in `stateObject.updateTrie` later as it's useless.